// d3-sankey-diagram/lib/link.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function () {
  var linkTitle = function linkTitle(d) {
    return defaultLinkTitle;
  };

  var path = (0, _linkPath2.default)();

  function sankeyLink(selection) {
    selection.each(function (d) {
      var link = _d2.default.select(this),
          transition = _d2.default.transition(link);

      if (!this._current) {
        // first time
        this._current = d;
        link.attr('d', path);
      }

      // Enter title
      var title = link.selectAll('title').data([0]);
      title.enter().append('title');

      // Update
      if (transition.attrTween) {
        transition.attrTween('d', tweenLink);
      } else {
        transition.attr('d', path);
        this._current = path;
      }

      link.select('title').text(linkTitle);
    });
  }

  // Store the displayed state in _current.
  // Then, interpolate from _current to the new state.
  // During the transition, _current is updated in-place by d3.interpolate.
  function tweenLink(b) {
    var i = _d2.default.interpolate(this._current, {
      x0: b.x0,
      y0: b.y0,
      x1: b.x1,
      y1: b.y1,
      dy: b.dy,
      r0: b.r0,
      r1: b.r1
    });
    this._current = i(0);
    return function (t) {
      var link = i(t);
      return path(i(t));
    };
  }

  sankeyLink.linkTitle = function (x) {
    if (!arguments.length) return linkTitle;
    linkTitle = _d2.default.functor(x);
    return sankeyLink;
  };

  return sankeyLink;
};

var _d = require('d3');

var _d2 = _interopRequireDefault(_d);

var _linkPath = require('./linkPath');

var _linkPath2 = _interopRequireDefault(_linkPath);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function defaultLinkTitle(d) {
  return d.source.id + ' → ' + d.target.id + '\n' + (d.data.type || '');
}
